lib/ansible/executor/task_executor.py

```diff
@@ -410,6 +410,13 @@ def _run_loop(self, items):
             results.append(res)
             del task_vars[loop_var]
 
+            # magic variables would be re-used in set_task_variable_override from previous iteration
+            # delete remote_user in case it's based on loop_var
+            # FIXME do we want to delete more than just remote_user?
+            for var in C.MAGIC_VARIABLE_MAPPING['remote_user']:
+                if var in task_vars:
+                    del task_vars[var]
+
```

Comment:
First, this is not the cannonical info for the user info anymore, you should use the connection plugin itself (which now includes vars specific to the plugin and not present in MAGIC_VARIABLE_MAPPING. 

Second, this should not be restricted to remote_user but 'all connection information' as port, host, etc can change per iteration see the `_set_plugin_options` and `_set_conneciton_options` functions for details.

Reply:
Ah, do you mean like this?
```diff
diff --git a/lib/ansible/executor/task_executor.py b/lib/ansible/executor/task_executor.py
index e3ee2547b5..96e0d462dd 100644
--- a/lib/ansible/executor/task_executor.py
+++ b/lib/ansible/executor/task_executor.py
@@ -410,10 +410,8 @@ class TaskExecutor:
             results.append(res)
             del task_vars[loop_var]
 
-            # magic variables would be re-used in set_task_variable_override from previous iteration
-            # delete remote_user in case it's based on loop_var
-            # FIXME do we want to delete more than just remote_user?
-            for var in C.MAGIC_VARIABLE_MAPPING['remote_user']:
+            # clear connection variables for next iteration
+            for var in C.config.get_plugin_vars('connection', self._connection._load_name):
                 if var in task_vars:
                     del task_vars[var]
```

Reply:
you probably need to add shell and become plugins also